frontend/src/views/Results.tsx

```diff
@@ -28,36 +31,65 @@ function GameResultsPage() {
   const location = useLocation<LocationState>();
 
   const [error, setError] = useState('');
-  const [loading, setLoading] = useState(false);
+  const [loading, setLoading] = useState(true);
 
-  const [players, setPlayers] = useState<Player[]>();
+  const [players, setPlayers] = useState<Player[]>([]);
   const [currentUser, setCurrentUser] = useState<User | null>(null);
-  const [room, setRoom] = useState<Room | null>(null);
+  const [host, setHost] = useState<User | null>(null);
+  const [roomId, setRoomId] = useState('');
 
   useEffect(() => {
-    if (checkLocationState(location, 'game', 'currentUser')) {
-      setPlayers(location.state.game.players);
-      setRoom(location.state.game.room);
+    if (checkLocationState(location, 'roomId', 'currentUser')) {
+      setRoomId(location.state.roomId);
       setCurrentUser(location.state.currentUser);
 
+      // Function that's called when playAgain is triggered
+      const playAgainAction = (game: Game) => {
+        disconnect()
+          .then(() => {
+            history.replace(`/game/lobby?room=${game.room.roomId}`, {
+              user: location.state.currentUser,
+              roomId: game.room.roomId,
+            });
+          });
+      };
+
       const subscribeCallback = (result: Message) => {
         const updatedGame: Game = JSON.parse(result.body);
 
+        // Update leaderboard with last second submissions
+        setPlayers(updatedGame.players);
+        // Set new host if the previous host refreshes or leaves
+        setHost(updatedGame.room.host);
+
         // Disconnect users from socket and then redirect them to the lobby page
         if (updatedGame.playAgain) {
-          disconnect()
-            .then(() => {
-              history.replace(`/game/lobby?room=${updatedGame.room.roomId}`, {
-                user: location.state.currentUser,
-                roomId: updatedGame.room.roomId,
-              });
-            })
-            .catch((err) => setError(err.message));
+          playAgainAction(updatedGame);
         }
       };
 
-      subscribe(routes(location.state.game.room.roomId).subscribe_game, subscribeCallback)
-        .catch((err) => setError(err.message));
+      /**
+       * Connect, subscribe, and then finally get the game details. Doing so in this order ensures
+       * that any late submissions are properly received (either through the socket update or
+       * through the get game request) and reflected on the leaderboard.
```

Comment:
So if a user submits their code right before the timer ends, the results page will initially display something, then will update through socket. I'm thinking that a good addition to make (doesn't have to be in this PR or sprint, just an idea) would be to have a field in the `Game` class that indicates a user's submission is currently being calculated, and if so, you could display something on the frontend that indicates as such, showing that the scores might change as the final submission comes in.

Reply:
I think this is a great idea - I'll add a card to the backlog for this (hopefully this type of thing doesn't show up too often but at some point it'll def be a good idea to make a fix)